### ISSUE TITLE
Serialisation Method and Keypaths (akin to Marshmallow Method/Function fields and attribute)

### ISSUE THREAD

**Issue by Alphasite** [samuelcolvin/pydantic#813]
# Feature Request

Please complete:
* OS: **mac 10.12**
* Python version `import sys; print(sys.version)`: **3.7.1 (default, Dec 20 2018, 13:24:33)**
* Pydantic version `import pydantic; print(pydantic.VERSION)`: **0.32.2**

**Please read the [docs](https://pydantic-docs.helpmanual.io/) and search through issues to
confirm your feature hasn't been asked for before, or already implemented.**

Where possible please include a self contained code snippet describing your feature request:

Similar to: https://marshmallow.readthedocs.io/en/stable/custom_fields.html#method-fields
and: https://github.com/marshmallow-code/marshmallow/issues/1315 

I would like to be able to provide a method to generate a field from an object, when ORM mode is used. 

Additionally I would like to be able to use a fully qualified 'dotted path' to reference nested fields (e.g. to flatten an orm object (obj.a.b.c -> obj.c)).

An example of what I would like to do:
```py
class User(SchemaBase):
    key_path: str
    method: str

    class Config:
        orm_mode = True


@dataclass
class Nested:
    test: str = "key serialiser!"


@dataclass
class UserModel:
    name: str
    nested: Nested = Nested()

user_model = UserModel("method")
user = User.from_orm(user_model)
print(user)
# > User key_path='key serialiser!' method='method serialiser!'
```

And an example (straw man) implementation of it, to be more concrete:

```py
"""
An example implementation and output from what i am trying to ask for;
see the bottom few lines for the actual expected input/output.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Type, Dict, Optional
from pydantic import BaseModel
from pydantic.utils import GetterDict

_sentinel = object()  # unique object used to signal cache misses


class Path:
    def __init__(self, path: str):
        self.path = path

    def resolve(self, obj: Any, default: Any, cls: Type):
        for attr_name in self.path.split("."):
            obj = getattr(obj, attr_name, _sentinel)

            if obj is _sentinel:
                return default

        return obj


class Function(Path):
    def __init__(self, path: str, schema_path: bool = True):
        super().__init__(path)
        self.schema_path = schema_path

    def resolve(self, obj: Any, default: Any, cls: Type):
        if self.schema_path:
            ref = super().resolve(cls, _sentinel, cls)
        else:
            ref = super().resolve(obj, _sentinel, cls)

        if ref is _sentinel:
            raise KeyError(f"Method: {self.path!r} not found")

        return ref(obj, default)


Resolvers = Dict[str, Path]


class ExtendedGetterDict(GetterDict):
    def __init__(self, obj: Any, resolvers: Resolvers, cls: Type):
        super().__init__(obj)
        self.cls = cls
        self.resolvers = resolvers

    def get(self, item: Any, default: Any) -> Any:
        """Add support for dump methods and key-paths"""

        if item in self.resolvers:
            return self.resolvers[item].resolve(self._obj, default, self.cls)

        return super().get(item, default)


class SchemaBase(BaseModel):
    """Add support for dump methods and key-paths

    Expect mappings in the form:
        __mapping__ = {
            "key_path": "nested.test",
            "method": get_value,
        }
    """

    __mapping__: Resolvers = {}
    __cached_mapping__: Optional[Resolvers] = None

    @classmethod
    def _decompose_class(cls: Type['Model'], obj: Any) -> GetterDict:
        return ExtendedGetterDict(obj, cls.__get_mapping__(), cls)

    @classmethod
    def __get_mapping__(cls) -> Resolvers:
        if not cls.__cached_mapping__:
            mapping = {}
            mapping.update(cls.__mapping__)

            # Union of own mapping and those of superclasses
            for base in cls.__bases__:
                if issubclass(base, SchemaBase):
                    mapping.update(base.__get_mapping__())

            cls.__cached_mapping__ = mapping

        return cls.__cached_mapping__


class User(SchemaBase):
    key_path: str
    method: str

    @staticmethod
    def get_value(object: UserModel, default):
        return object.name + " serialiser!"

    __mapping__ = {
        "key_path": Path("nested.test"),
        "method": Function("get_value"),
    }

    class Config:
        orm_mode = True


@dataclass
class Nested:
    test: str = "key serialiser!"


@dataclass
class UserModel:
    name: str
    nested: Nested = Nested()


user_model = UserModel("method")
user = User.from_orm(user_model)
print(user)
# > User key_path='key serialiser!' method='method serialiser!'
```

I would ideally like to integrate the path (for example) into the Schema Object.

**Comment by samuelcolvin** [comment]
> An example of what I would like to do:

Maybe I'm being dumb, but I can't see what's changed / what the behaviour your want is?

**Comment by Alphasite** [comment]
Sorry, i guess i didn't explain it very well. 

I think i confused two things by mixing my two requests (and proving an example implementation); so let me separate them a little. 

# Nested ORM Mapping 

The shape of the first problem is that the ORM model(s) and the external schema have different 'shapes' as it were. The external schema has nested fields (e.g. `User.company.name` is represented as just `company_name` in external schema. 

e.g. 
```
class Company:
     name: str

class User:
     name: str
     company: Company

class UserResponseSchema(SchemaBase):
     name: str
     company: str = Schema(attribute="company.name")
```

# ORM -> Schema Model Conversion Functions 

Given an input, i would like to be able to run a function to extract arbitrary data from the form object. Currently `@validation` can be used to do emulate this when deserialising, but there is no equivalent for ORM objects (as I understand it).

```
class User:
     first_name: str
     second_name: str

class UserResponseSchema(SchemaBase):
     name: str = Schema(serialisation_function="serialise_name")

     @static_function
     def serialise_name(input: User) -> str:
           return f"{input.first_name} {input.second_name}"
```

**Comment by Alphasite** [comment]
i extracted that from the big block of code below and removed some scaffolding code, but didn't check to see if it actually made sense in isolation.

**Comment by dmontagu** [comment]
1. In response to your `Nested ORM Mapping` point, this makes sense to me as a real problem. I think the simplest change to pydantic that could enable this would be to:
    1. Make `BaseModel._decompose_class` non-private so it could be overridden -- this would enable you to manually create the dict you want to produce from the ORM object (instead of just using a `GetterDict`)
    2. Ideally, make it easier to override `GetterDict.get` for specific field names. I'm not sure what this API should look like, but it could help make the overridden `BaseModel._decompose_class` shorter if you just want special treatment of one or two fields.

    Alternatively, @samuelcolvin might it be possible to accomplish this by means of the new root_validator functionality for models without a custom root type?


1. In response to your `ORM -> Schema Model Conversion Functions` point, pydantic currently avoids lazy evaluation, so, as far as I am aware, there is nothing analogous to specifying a `serialisation_function`. Also, I think that needing to check for a `serialisation_function` when serialising every attribute might come with a macroscopic performance penalty.

    I personally prefer it this way as it makes it easier to reason about when computations will be done, and prevents confusion about differences between the serialized and unserialized values. But to each their own.

**Comment by Alphasite** [comment]
I was specifically thinking of the validate_always example from the docs, where they use the validate to insert a computed default value into a field. 

And adding an extension point, to let me implement this would also be a workable solution.

**Comment by samuelcolvin** [comment]
`_decompose_class` can already be override. I'm completely opposed to making it properly public, the reason we didn't before was to avoid polluting the field namespace. It's setup the way it is **precisely** to allow custom implementations, I guess we should make it public and add docs for this.

Regarding `Schema(attribute='company.name')` I'm not opposed to it, I think rust's serde library has this and I've used it in the past. But it's not going to be done in the near future. You might find root validators (implemented in #817) useful as a substitute, although for them to work, `company.name` would have to be non-required.

**Comment by samuelcolvin** [comment]
with `_decompose_class` and `config.getter_dict` I think this is covered.
